src/main/scala/kinesis/mock/api/PutRecordsRequest.scala

```diff
@@ -54,63 +56,58 @@ final case class PutRecordsRequest(
             }
         )
         .map { case (stream, recs) =>
-          val grouped = recs
-            .groupBy { case (shard, records, _) =>
-              (shard, records)
-            }
-            .map { case ((shard, records), list) =>
-              (shard, records) ->
-                list.map(_._3).zipWithIndex.map { case (entry, index) =>
-                  val seqNo = SequenceNumber.create(
-                    shard.createdAtTimestamp,
-                    shard.shardId.index,
-                    None,
-                    Some(records.length + index),
-                    Some(now)
-                  )
-                  (
-                    KinesisRecord(
-                      now,
-                      entry.data,
-                      stream.encryptionType,
-                      entry.partitionKey,
-                      seqNo
-                    ),
-                    PutRecordsResultEntry(
-                      None,
-                      None,
-                      Some(seqNo),
-                      Some(shard.shardId.shardId)
-                    )
-                  )
-                }
-            }
-            .toVector
+          val asRecords = PutRecordsRequest
+            .getIndexByShard(recs)
+            .map { case (shard, records, entry, index) =>
+              val seqNo = SequenceNumber.create(
+                shard.createdAtTimestamp,
+                shard.shardId.index,
+                None,
+                Some(records.length + index),
+                Some(now)
+              )
 
-          val newShards = grouped.map {
-            case ((shard, currentRecords), recordsToAdd) =>
               (
                 shard,
-                (
-                  currentRecords ++ recordsToAdd.map(_._1),
-                  recordsToAdd.map(_._2)
+                records,
+                KinesisRecord(
+                  now,
+                  entry.data,
+                  stream.encryptionType,
+                  entry.partitionKey,
+                  seqNo
+                ),
+                PutRecordsResultEntry(
+                  None,
+                  None,
+                  Some(seqNo),
+                  Some(shard.shardId.shardId)
                 )
               )
-          }
+            }
+
+          val newShards = asRecords
+            .groupBy { case (shard, currentRecords, _, _) =>
```

Comment:
The groupBy here is still needed for calculating the new shard data after the put is complete. It is safe since it is not used in the response itself.